File: day12/aoc2025_12.scala
def main(args: Array[String]): Unit = {
    val blocks = scala.io.Source
        .fromFile("src/main/scala/input12")
        .mkString
        .split("\n\n")
    val (regionBlocks, shapeBlocks) = blocks.partition(_.contains("x"))
    val shapeAreas = shapeBlocks.map(_.count(_ == '#'))
    val regions = regionBlocks.flatMap(_.split("\n"))
    val count = regions.count { line =>
        val parts = line.split(":")
        val regionSize = parts(0).split("x").map(_.toInt).product
        val counts = parts(1).trim.split("\\s+").map(_.toInt)
        val totalShapeArea = counts.zip(shapeAreas).map { case (n, a) => n * a }.sum
        totalShapeArea <= regionSize
    }
    println("Part 1: " + count)
}

File: day12/aoc2025_12.ts
import { readFileSync } from "node:fs";

const blocks = readFileSync("day12/input12", "utf-8")
    .trim()
    .split("\n\n");

const regionBlocks = blocks.filter(b => b.includes("x"));
const shapeBlocks = blocks.filter(b => !b.includes("x"));
const shapeAreas = shapeBlocks.map(b => b.split("").filter(c => c === "#").length);
const regions = regionBlocks.flatMap(b => b.split("\n"));
const count = regions.filter(line => {
    const [dimPart, countsPart] = line.split(":");
    const regionSize = dimPart
        .split("x")
        .map(Number)
        .reduce((a, b) => a * b, 1);
    const counts = countsPart.trim().split(/\s+/).map(Number);
    const totalShapeArea = counts.reduce((sum, n, i) => sum + n * shapeAreas[i], 0);
    return totalShapeArea <= regionSize;
}).length;

console.log("Part 1:", count);
